public/app/plugins/panel/multistat/module.ts

```diff
@@ -0,0 +1,135 @@
+    super($scope, $injector);
+
+    _.defaults(this.panel, defaults.panelDefaults);
+
+    this.events.on('data-received', this.onDataReceived.bind(this));
+    this.events.on('data-error', this.onDataError.bind(this));
+    this.events.on('data-snapshot-load', this.onDataReceived.bind(this));
+    this.events.on('init-edit-mode', this.onInitEditMode.bind(this));
+  }
+
+  onInitEditMode() {
+    this.fontSizes = ['20%', '30%', '50%', '70%', '80%', '100%', '110%', '120%', '150%', '170%', '200%'];
+    this.addEditorTab('Options', 'public/app/plugins/panel/multistat/options.html', 2);
+    this.addEditorTab('Thresholds', 'public/app/plugins/panel/multistat/thresholds.html', 3);
+    this.unitFormats = kbn.getUnitFormats();
+  }
+
+  onDataReceived(dataList) {
+    if (dataList.length > 0 && dataList[0].type === 'table') {
+      this.dataType = 'table';
+      if (dataList[0].rows && !dataList[0].rows.length) {
+        return this.onDataError('No data');
+      }
+      this.setTableColumnToSensibleDefault(dataList[0]);
+      this.data = convertTableDataToMultistat(dataList, this.panel);
+    } else {
+      this.dataType = 'timeseries';
+      this.data = convertTSDataToMultistat(dataList, this.panel);
+      // this.setValues();
+    }
+    // this.data = data;
+    this.render();
+  }
+
+  onDataError(err) {
+    this.onDataReceived([]);
+  }
+
+  setTableColumnToSensibleDefault(tableData) {
+    const columnNames = {};
+
+    tableData.columns.forEach((column, columnIndex) => {
+      columnNames[columnIndex] = column.text;
+    });
+
+    this.tableColumnOptions = columnNames;
+    if (
+      _.find(tableData.columns, ['text', this.panel.tableColumnValue]) &&
+      _.find(tableData.columns, ['text', this.panel.tableColumnLabel])
+    ) {
+      return;
+    }
+
+    if (tableData.columns.length === 1) {
+      this.panel.tableColumnValue = tableData.columns[0].text;
+    } else {
+      const notTimeColumns = _.filter(tableData.columns, col => col.type !== 'time');
+      this.panel.tableColumnValue = _.last(notTimeColumns).text;
+      this.panel.tableColumnLabel = _.first(notTimeColumns).text;
+    }
+  }
+
+  setValueMapping(data) {}
+
+  setUnitFormat(subItem) {
+    this.panel.format = subItem.value;
+    this.refresh();
+  }
+
+  onThresholdChange = newThresholds => {
+    this.panel.thresholds = newThresholds;
+    this.render();
+  };
+
+  link(scope, elem, attrs, ctrl) {
+    const multistatElem = elem.find('.multistat-panel');
+
+    function render() {
+      if (!ctrl.data) {
+        return;
+      }
+
+      const width = multistatElem.width();
+      const height = multistatElem.height();
+      scope.size = { w: width, h: height };
+      renderMultiStatComponent();
+    }
+
+    function renderMultiStatComponent() {
+      const multistatProps = {
+        stats: ctrl.data,
+        options: ctrl.panel,
+        size: scope.size,
+      };
+      const multistatReactElem = React.createElement(MultiStat, multistatProps);
```

Comment:
Is this needed every render? Can you check if MutiStat is demounted & mounted again every render? Maybe there is a more efficient way to re-render the panel by just having access to the component instance and calling forceUpdate()  ?

Reply:
Yes, I think it makes sense.

Reply:
@torkelo investigating this, but have a problem. We should update props every render (since data, panel size or other options is changed).  But, seems, there's no way to update props for the element, created by `React.createElement()`. On the other hand, [render()](https://reactjs.org/docs/react-dom.html#render) docs says:
>If the React element was previously rendered into container, this will perform an update on it and only mutate the DOM as necessary to reflect the latest React element.

So maybe it's not a big overhead. Any thoughts?

Reply:
Found another potential problem. When user does actions which change props and trigger component re-rendering, it causes re-rendering of __all__ child components. For example: [here](https://vimeo.com/281973954) I'm changing threshold color and this action should update only related components (`SingleStat` component get color as a property, so if it has not been changed it shouldn't be re-rendered), but all children are updating.

Reply:
Seems, `React.PureComponent` solves this problem.